src/network/network_server.cpp

```diff
@@ -652,8 +652,22 @@ NetworkRecvStatus ServerNetworkGameSocketHandler::SendMap()
 				return NETWORK_RECV_STATUS_CONN_LOST;
 
 			case SPS_ALL_SENT:
-				/* All are sent, increase the sent_packets */
-				if (has_packets) sent_packets *= 2;
+				/*
+				 * All are sent, increase the sent_packets. However, do not increase the sent_packets so
+				 * far that it would overflow and become 0. With threaded saving the sent_packets can get
+				 * huge quickly if the saving releases a single packet every tick. If that packet could
+				 * then be put in the network buffer immediately, all packets would be sent and the limit
+				 * would be doubled. This regardless whether that amount of packets was actually sent.
+				 * When the savegame is dumped onto the connection immediately, like when there is no
+				 * threading, then the network buffer could be filled quite quickly causing blocks on
+				 * certain OSes even though we request the writing to be non-blocking. As such this limit
+				 * on the number of packets that can be sent each tick has been introduced.
+				 * When using relatively fast threaded saves and a slow network connection, then the
+				 * kick back would be triggered relatively quickly.
+				 */
```

Comment:
Although make sure we don't overflow is always a good thing, it made me wonder based on your comment that we are not actually fixing the wrong thing?
Shouldn't we only double if the amount we queued was equal to `sent_packets` to start with?

The reason is another scenario: the savegame compression is slow at start, so `sent_packets` is getting really large. All of a sudden the compression is really fast, and we have a huge queue  which we keep on filling every tick till the `SPS_PARTLY_SENT` as made it to a sane value again.

Possibly fixing that, also prevents the need for such a large comment, as code-wise that would make much more sense (don't double `sent_packets` if we sent fewer than `sent_packets` anyway).

Reply:
It is certain there are other issues with it, and even the question whether it is really needed, and if so whether it should be improved. Now it's quite binary, and steering the number of packets to some reasonable amount is difficult. Practically you probably want to keep some packets in the buffer but at a rate that it's not drying up and also not too many packets are in the queue. So it should know how many packets it put in the queue and how many are remaining after sending, and based on that determine the actions.
Then of course there is the question whether the whole limiting is actually (still) needed.